File: dlna-dms/src/main/java/com/android/cast/dlna/dms/IMediaContentDao.java
package com.android.cast.dlna.dms;

import android.content.Context;
import android.database.Cursor;
import android.provider.MediaStore;

import androidx.annotation.NonNull;

import org.fourthline.cling.support.model.PersonWithRole;
import org.fourthline.cling.support.model.Res;
import org.fourthline.cling.support.model.item.ImageItem;
import org.fourthline.cling.support.model.item.Item;
import org.fourthline.cling.support.model.item.Movie;
import org.fourthline.cling.support.model.item.MusicTrack;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

interface IMediaContentDao {
    @NonNull
    List<Item> getImageItems(@NonNull Context context);

    @NonNull
    List<Item> getAudioItems(@NonNull Context context);

    @NonNull
    List<Item> getVideoItems(@NonNull Context context);

    // -------------------------------------------------------
    // ---- Implement
    // -------------------------------------------------------
    final class MediaContentDao implements IMediaContentDao {

        private static final String[] CONTENT_IMAGE_COLUMNS = {MediaStore.MediaColumns._ID, MediaStore.MediaColumns.TITLE, MediaStore.MediaColumns.ARTIST,
                MediaStore.MediaColumns.DATA, MediaStore.MediaColumns.MIME_TYPE, MediaStore.MediaColumns.SIZE};
        private static final String[] CONTENT_AUDIO_COLUMNS = {MediaStore.MediaColumns._ID, MediaStore.MediaColumns.TITLE, MediaStore.MediaColumns.ARTIST,
                MediaStore.MediaColumns.DATA, MediaStore.MediaColumns.MIME_TYPE, MediaStore.MediaColumns.SIZE, MediaStore.MediaColumns.ALBUM};
        private static final String[] CONTENT_VIDEO_COLUMNS = {MediaStore.MediaColumns._ID, MediaStore.MediaColumns.TITLE, MediaStore.MediaColumns.ARTIST,
                MediaStore.MediaColumns.DATA, MediaStore.MediaColumns.MIME_TYPE, MediaStore.MediaColumns.SIZE, MediaStore.MediaColumns.RESOLUTION};

        private final String mBaseUrl;

        public MediaContentDao(String baseUrl) {
            mBaseUrl = baseUrl;
        }

        @Override
        @NonNull
        public List<Item> getImageItems(@NonNull Context context) {
            List<Item> items = new ArrayList<>();
            try (Cursor cursor = context.getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, CONTENT_IMAGE_COLUMNS, null, null, null)) {
                if (cursor == null) return items;
                cursor.moveToFirst();
                while (cursor.moveToNext()) {
                    String id = String.valueOf(cursor.getInt(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[0])));
                    String title = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[1]));
                    String creator = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[2]));
                    String data = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[3]));
                    String mimeType = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[4]));
                    long size = cursor.getLong(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[5]));
                    String url = mBaseUrl + File.separator + data;
                    Res res = new Res(mimeType, size, "", null, url);
                    ImageItem imageItem = new ImageItem(id, MediaItem.IMAGE_ID, title, creator, res);
                    items.add(imageItem);
                }
            }
            return items;
        }

        @NonNull
        public List<Item> getAudioItems(@NonNull Context context) {
            List<Item> items = new ArrayList<>();
            try (Cursor cursor = context.getContentResolver().query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, CONTENT_AUDIO_COLUMNS, null, null, null)) {
                if (cursor == null) return items;
                cursor.moveToFirst();
                while (cursor.moveToNext()) {
                    String id = String.valueOf(cursor.getInt(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[0])));
                    String title = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[1]));
                    String creator = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[2]));
                    String data = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[3]));
                    String mimeType = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[4]));
                    long size = cursor.getLong(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[5]));
                    // long duration = cursor.getLong(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[6]));
                    // String durationStr = ModelUtil.toTimeString(duration / 1000);
                    String album = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[7]));
                    String url = mBaseUrl + File.separator + data;
                    Res res = new Res(mimeType, size, "", null, url);
                    MusicTrack musicTrack = new MusicTrack(id, MediaItem.AUDIO_ID, title, creator, album, new PersonWithRole(creator), res);
                    items.add(musicTrack);
                }
            }

            return items;
        }

        @NonNull
        public List<Item> getVideoItems(@NonNull Context context) {
            List<Item> items = new ArrayList<>();
            try (Cursor cur = context.getContentResolver().query(MediaStore.Video.Media.EXTERNAL_CONTENT_URI, CONTENT_VIDEO_COLUMNS, null, null, null)) {
                if (cur == null) return items;
                cur.moveToFirst();
                while (cur.moveToNext()) {
                    String id = String.valueOf(cur.getInt(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[0])));
                    String title = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[1]));
                    String creator = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[2]));
                    String data = cur.getString(cur.getColumnIndexOrThrow(CONTENT_VIDEO_COLUMNS[3]));
                    String mimeType = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[4]));
                    long size = cur.getLong(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[5]));
                    // long duration = cur.getLong(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[6]));
                    // String durationStr = ModelUtil.toTimeString(duration / 1000);
                    String resolution = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[7]));
                    String url = mBaseUrl + File.separator + data;
                    Res res = new Res(mimeType, size, "", null, url);
                    res.setResolution(resolution);
                    Movie movie = new Movie(id, MediaItem.VIDEO_ID, title, creator, res);
                    items.add(movie);
                }
            }
            return items;
        }
    }
}

File: dlna-dms/src/main/java/com/android/cast/dlna/dms/IMediaContentDao.kt
package com.android.cast.dlna.dms

import android.annotation.SuppressLint
import android.content.Context
import android.provider.MediaStore.*
import android.provider.MediaStore.Images.Media
import com.android.cast.dlna.core.ContentType.*
import org.fourthline.cling.support.model.PersonWithRole
import org.fourthline.cling.support.model.Res
import org.fourthline.cling.support.model.item.ImageItem
import org.fourthline.cling.support.model.item.Item
import org.fourthline.cling.support.model.item.Movie
import org.fourthline.cling.support.model.item.MusicTrack

// ------------------------------------------------------------
// ---- MediaContent
// ------------------------------------------------------------
interface IMediaContentDao {

    companion object {
        private val CONTENT_IMAGE_COLUMNS = arrayOf(
            MediaColumns._ID,
            MediaColumns.TITLE,
            MediaColumns.DATA,
            MediaColumns.MIME_TYPE,
            MediaColumns.SIZE
        )
        private val CONTENT_AUDIO_COLUMNS = arrayOf(
            MediaColumns._ID,
            MediaColumns.TITLE,
            MediaColumns.ARTIST,
            MediaColumns.DATA,
            MediaColumns.MIME_TYPE,
            MediaColumns.SIZE,
            MediaColumns.ALBUM
        )
        private val CONTENT_VIDEO_COLUMNS = arrayOf(
            MediaColumns._ID,
            MediaColumns.TITLE,
            MediaColumns.ARTIST,
            MediaColumns.DATA,
            MediaColumns.MIME_TYPE,
            MediaColumns.SIZE,
            MediaColumns.RESOLUTION
        )
    }

    fun getImageItems(context: Context): List<Item?>
    fun getAudioItems(context: Context): List<Item?>
    fun getVideoItems(context: Context): List<Item?>

    // -------------------------------------------------------
    // ---- Implement
    // -------------------------------------------------------
    //TODO: check range issue.
    @SuppressLint("Range")
    class MediaContentDao(private val mBaseUrl: String) : IMediaContentDao {
        override fun getImageItems(context: Context): List<Item> {
            val items: MutableList<Item> = ArrayList()
            context.contentResolver.query(Media.EXTERNAL_CONTENT_URI, CONTENT_IMAGE_COLUMNS, null, null, null)
                .use { cursor ->
                    if (cursor == null) return items
                    cursor.moveToFirst()
                    while (cursor.moveToNext()) {
                        val id = cursor.getInt(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[0])).toString()
                        val title = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[1]))
                        val data = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[2]))
                        val mimeType = cursor.getString(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[3]))
                        val size = cursor.getLong(cursor.getColumnIndex(CONTENT_IMAGE_COLUMNS[4]))
                        val url = mBaseUrl + data
                        val res = Res(mimeType, size, "", null, url)
                        val imageItem = ImageItem(id, IMAGE.id, title, "", res)
                        items.add(imageItem)
                    }
                }
            return items
        }

        override fun getAudioItems(context: Context): List<Item> {
            val items: MutableList<Item> = ArrayList()
            context.contentResolver.query(Audio.Media.EXTERNAL_CONTENT_URI, CONTENT_AUDIO_COLUMNS, null, null, null)
                .use { cursor ->
                    if (cursor == null) return items
                    cursor.moveToFirst()
                    while (cursor.moveToNext()) {
                        val id = cursor.getInt(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[0])).toString()
                        val title = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[1]))
                        val creator = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[2]))
                        val data = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[3]))
                        val mimeType = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[4]))
                        val size = cursor.getLong(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[5]))
                        // long duration = cursor.getLong(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[6]));
                        // String durationStr = ModelUtil.toTimeString(duration / 1000);
                        val album = cursor.getString(cursor.getColumnIndex(CONTENT_AUDIO_COLUMNS[6]))
                        val url = mBaseUrl + data
                        val res = Res(mimeType, size, "", null, url)
                        val musicTrack = MusicTrack(id, AUDIO.id, title, creator, album, PersonWithRole(creator), res)
                        items.add(musicTrack)
                    }
                }
            return items
        }

        override fun getVideoItems(context: Context): List<Item> {
            val items: MutableList<Item> = ArrayList()
            context.contentResolver.query(Video.Media.EXTERNAL_CONTENT_URI, CONTENT_VIDEO_COLUMNS, null, null, null)
                .use { cur ->
                    if (cur == null) return items
                    cur.moveToFirst()
                    while (cur.moveToNext()) {
                        val id = cur.getInt(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[0])).toString()
                        val title = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[1]))
                        val creator = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[2]))
                        val data = cur.getString(cur.getColumnIndexOrThrow(CONTENT_VIDEO_COLUMNS[3]))
                        val mimeType = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[4]))
                        val size = cur.getLong(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[5]))
                        // long duration = cur.getLong(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[6]));
                        // String durationStr = ModelUtil.toTimeString(duration / 1000);
                        val resolution = cur.getString(cur.getColumnIndex(CONTENT_VIDEO_COLUMNS[6]))
                        val url = mBaseUrl + data
                        val res = Res(mimeType, size, "", null, url)
                        res.resolution = resolution
                        val movie = Movie(id, VIDEO.id, title, creator, res)
                        items.add(movie)
                    }
                }
            return items
        }

    }
}